### ISSUE TITLE
[CSX] Error during BS script execution (fCCD related?)

### ISSUE THREAD

**Issue by cmazzoli** [bluesky/bluesky#1159]
Several problems seen yesterday are reported below.
Please advise.

I report the bad script like but with a properly defined function, it gave the same error..
(like:
```
def testfunc(): 
    ...:     while(True): 
    ...:         yield from asc([sclr],pgm.energy, 920, 950, 31) 
    ...:         nap(600) 

btw nap = bps.sleep
)
''''
|        29 | 01:48:25.1 |             947.854 |      13040 |                                                                                                                                       
|        30 | 01:48:30.9 |             948.927 |      10864 |                                                                                                                                       
|        31 | 01:48:36.7 |             950.190 |      11284 |                                                                                                                                       
+-----------+------------+---------------------+------------+
generator scan ['3fb1efbe'] (scan num: 114291)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
~/Beamline/Commissioning/2019_03/exit_arm.py in <module>
      1 while(True):
      2     RE(asc([sclr],pgm.energy, 920, 950, 31))
----> 3     RE(nap(600))
      4 

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/run_engine.py in __call__(self, *args, **metadata_kw)
    726         self._metadata_per_call.update(metadata_kw)
    727 
--> 728         gen = ensure_generator(plan)
    729         for wrapper_func in self.preprocessors:
    730             gen = wrapper_func(gen)

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/utils.py in ensure_generator(plan)
    110     if isinstance(plan, Msg):
    111         return single_gen(plan)
--> 112     gen = iter(plan)  # no-op on generators; needed for classes
    113     if not isinstance(gen, PLAN_TYPES):
    114         # If plan does not support .send, we must wrap it in a generator.

TypeError: 'NoneType' object is not iterable


''''

This was used to work...
''''
In [29]: RE(spiral_continuous(dets,nanop.bx,nanop.bz,-2.2965,-0.8338,50,0.0005,0.95))                                                                                                               
Transient Scan ID: 114371     Time: 2019/03/11 01:54:24
Persistent Unique Scan ID: '126f1f4d-549c-488a-9c1f-0cfdcad43c98'
New stream: 'baseline'
New stream: 'primary'                                                                                                                                                                               
+-----------+------------+------------+------------+-------------------+-------------------+-------------------+-------------------+-------------------+                                            
|   seq_num |       time |   nanop_bx |   nanop_bz | fccd_stats1_total | fccd_stats2_total | fccd_stats3_total | fccd_stats4_total | fccd_stats5_total |
+-----------+------------+------------+------------+-------------------+-------------------+-------------------+-------------------+-------------------+
+-----------+------------+------------+------------+-------------------+-------------------+-------------------+-------------------+-------------------+
generator spiral_continuous ['126f1f4d'] (scan num: 114371)
Moving back to first point position.
[E 01:54:34.982 run_engine:1264] Run aborted                                                                                                                                                        
    Traceback (most recent call last):                                                                                                                                                              
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/run_engine.py", line 1153, in _run
        msg = self._plan_stack[-1].send(resp)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 1294, in __call__
        return (yield from plan)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 1149, in baseline_wrapper
        return (yield from plan_mutator(plan, insert_baseline))
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 169, in plan_mutator
        raise ex
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 122, in plan_mutator
        msg = plan_stack[-1].send(ret)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 790, in monitor_during_wrapper
        return (yield from plan2)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 169, in plan_mutator
        raise ex
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 122, in plan_mutator
        msg = plan_stack[-1].send(ret)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 169, in plan_mutator
        raise ex
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 122, in plan_mutator
        msg = plan_stack[-1].send(ret)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 848, in fly_during_wrapper
        return (yield from plan2)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 169, in plan_mutator
        raise ex
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 122, in plan_mutator
        msg = plan_stack[-1].send(ret)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 169, in plan_mutator
        raise ex
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 122, in plan_mutator
        msg = plan_stack[-1].send(ret)
      File "/epics/xf/23id/ipython/profile_collection/startup/csx1/plans/nanop.py", line 112, in spiral_continuous
        return (yield from plan_steps())
      File "/epics/xf/23id/ipython/profile_collection/startup/csx1/plans/nanop.py", line 107, in plan_steps
        yield from cont_sp_plan
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/plans.py", line 750, in scan_nd
        return (yield from inner_scan_nd())
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/utils.py", line 1011, in dec_inner
        return (yield from plan)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 939, in stage_wrapper
        return (yield from finalize_wrapper(inner(), unstage_devices()))
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 500, in finalize_wrapper
        ret = yield from plan
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 937, in inner
        return (yield from plan)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/utils.py", line 1011, in dec_inner
        return (yield from plan)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 326, in run_wrapper
        else_plan=close_run)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 562, in contingency_wrapper
        ret = yield from plan
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/plans.py", line 748, in inner_scan_nd
        yield from per_step(detectors, step, pos_cache)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/plan_stubs.py", line 952, in one_nd_step
        yield from trigger_and_read(list(detectors) + list(motors))
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/plan_stubs.py", line 802, in trigger_and_read
        rewindable))
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 680, in rewindable_wrapper
        return (yield from plan)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/plan_stubs.py", line 798, in inner_trigger_and_read
        yield from save()
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/plan_stubs.py", line 57, in save
        return (yield Msg('save'))
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 194, in plan_mutator
        inner_ret = yield msg
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 194, in plan_mutator
        inner_ret = yield msg
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 194, in plan_mutator
        inner_ret = yield msg
      [Previous line repeated 1 more times]
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/run_engine.py", line 1211, in _run
        new_response = yield from coro(msg)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/run_engine.py", line 1687, in _save
        yield from self.emit(DocumentNames.descriptor, doc)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/asyncio/coroutines.py", line 212, in coro
        res = func(*args, **kw)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/run_engine.py", line 2324, in emit
        self.dispatcher.process(name, doc)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/run_engine.py", line 2344, in process
        exceptions = self.cb_registry.process(name, name.name, doc)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/utils.py", line 334, in process
        func(*args, **kwargs)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/utils.py", line 415, in __call__
        return mtd(*args, **kwargs)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/callbacks/best_effort.py", line 93, in __call__
        super().__call__(name, doc)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/callbacks/core.py", line 17, in __call__
        return getattr(self, name)(doc)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/callbacks/best_effort.py", line 341, in descriptor
        live_scatter('start', self._start_doc)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/callbacks/core.py", line 17, in __call__
        return getattr(self, name)(doc)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/callbacks/mpl_plotting.py", line 225, in start
        norm=self._norm, cmap=self.cmap, **self.kwargs)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/matplotlib/__init__.py", line 1805, in inner
        return func(ax, *args, **kwargs)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/matplotlib/axes/_axes.py", line 4195, in scatter
        isinstance(c[0], str))):
    IndexError: list index out of range
[E 01:54:34.982 run_engine:1264] Run aborted
    Traceback (most recent call last):
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/run_engine.py", line 1153, in _run
        msg = self._plan_stack[-1].send(resp)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 1294, in __call__
        return (yield from plan)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 1149, in baseline_wrapper
        return (yield from plan_mutator(plan, insert_baseline))
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 169, in plan_mutator
        raise ex
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 122, in plan_mutator
        msg = plan_stack[-1].send(ret)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 790, in monitor_during_wrapper
        return (yield from plan2)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 169, in plan_mutator
        raise ex
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 122, in plan_mutator
        msg = plan_stack[-1].send(ret)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 169, in plan_mutator
        raise ex
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 122, in plan_mutator
        msg = plan_stack[-1].send(ret)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 848, in fly_during_wrapper
        return (yield from plan2)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 169, in plan_mutator
        raise ex
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 122, in plan_mutator
        msg = plan_stack[-1].send(ret)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 169, in plan_mutator
        raise ex
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 122, in plan_mutator
        msg = plan_stack[-1].send(ret)
      File "/epics/xf/23id/ipython/profile_collection/startup/csx1/plans/nanop.py", line 112, in spiral_continuous
        return (yield from plan_steps())
      File "/epics/xf/23id/ipython/profile_collection/startup/csx1/plans/nanop.py", line 107, in plan_steps
        yield from cont_sp_plan
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/plans.py", line 750, in scan_nd
        return (yield from inner_scan_nd())
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/utils.py", line 1011, in dec_inner
        return (yield from plan)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 939, in stage_wrapper
        return (yield from finalize_wrapper(inner(), unstage_devices()))
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 500, in finalize_wrapper
        ret = yield from plan
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 937, in inner
        return (yield from plan)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/utils.py", line 1011, in dec_inner
        return (yield from plan)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 326, in run_wrapper
        else_plan=close_run)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 562, in contingency_wrapper
        ret = yield from plan
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/plans.py", line 748, in inner_scan_nd
        yield from per_step(detectors, step, pos_cache)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/plan_stubs.py", line 952, in one_nd_step
        yield from trigger_and_read(list(detectors) + list(motors))
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/plan_stubs.py", line 802, in trigger_and_read
        rewindable))
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 680, in rewindable_wrapper
        return (yield from plan)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/plan_stubs.py", line 798, in inner_trigger_and_read
        yield from save()
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/plan_stubs.py", line 57, in save
        return (yield Msg('save'))
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 194, in plan_mutator
        inner_ret = yield msg
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 194, in plan_mutator
        inner_ret = yield msg
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 194, in plan_mutator
        inner_ret = yield msg
      [Previous line repeated 1 more times]
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/run_engine.py", line 1211, in _run
        new_response = yield from coro(msg)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/run_engine.py", line 1687, in _save
        yield from self.emit(DocumentNames.descriptor, doc)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/asyncio/coroutines.py", line 212, in coro
        res = func(*args, **kw)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/run_engine.py", line 2324, in emit
        self.dispatcher.process(name, doc)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/run_engine.py", line 2344, in process
        exceptions = self.cb_registry.process(name, name.name, doc)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/utils.py", line 334, in process
        func(*args, **kwargs)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/utils.py", line 415, in __call__
        return mtd(*args, **kwargs)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/callbacks/best_effort.py", line 93, in __call__
        super().__call__(name, doc)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/callbacks/core.py", line 17, in __call__
        return getattr(self, name)(doc)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/callbacks/best_effort.py", line 341, in descriptor
        live_scatter('start', self._start_doc)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/callbacks/core.py", line 17, in __call__
        return getattr(self, name)(doc)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/callbacks/mpl_plotting.py", line 225, in start
        norm=self._norm, cmap=self.cmap, **self.kwargs)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/matplotlib/__init__.py", line 1805, in inner
        return func(ax, *args, **kwargs)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/matplotlib/axes/_axes.py", line 4195, in scatter
        isinstance(c[0], str))):
    IndexError: list index out of range
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
~/Beamline/ScienceComm/2019_03_ReconsTeam/Ptychography_routine.py in <module>
----> 1 RE(spiral_continuous(dets,nanop.bx,nanop.bz,-2.2965,-0.8338,50,0.0005,0.95))

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/run_engine.py in __call__(self, *args, **metadata_kw)
    754                     # it (unless it is a canceled error)
    755                     if exc is not None:
--> 756                         raise exc
    757 
    758             if self._interrupted:

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/run_engine.py in _run(self)
   1263             self._reason = str(err)
   1264             self.log.exception("Run aborted")
-> 1265             raise err
   1266         finally:
   1267             # Some done_callbacks may still be alive in other threads.

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/run_engine.py in _run(self)
   1151                     else:
   1152                         try:
-> 1153                             msg = self._plan_stack[-1].send(resp)
   1154                         # We have exhausted the top generator
   1155                         except StopIteration:

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in __call__(self, plan)
   1292         plan = monitor_during_wrapper(plan, self.monitors)
   1293         plan = baseline_wrapper(plan, self.baseline)
-> 1294         return (yield from plan)

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in baseline_wrapper(plan, devices, name)
   1147         return (yield from plan)
   1148     else:
-> 1149         return (yield from plan_mutator(plan, insert_baseline))
   1150 
   1151 

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in plan_mutator(plan, msg_proc)
    167                     continue
    168                 else:
--> 169                     raise ex
    170         # if inserting / mutating, put new generator on the stack
    171         # and replace the current msg with the first element from the

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in plan_mutator(plan, msg_proc)
    120             ret = result_stack.pop()
    121             try:
--> 122                 msg = plan_stack[-1].send(ret)
    123             except StopIteration as e:
    124                 # discard the exhausted generator

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in monitor_during_wrapper(plan, signals)
    788     plan1 = plan_mutator(plan, insert_after_open)
    789     plan2 = plan_mutator(plan1, insert_before_close)
--> 790     return (yield from plan2)
    791 
    792 

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in plan_mutator(plan, msg_proc)
    167                     continue
    168                 else:
--> 169                     raise ex
    170         # if inserting / mutating, put new generator on the stack
    171         # and replace the current msg with the first element from the

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in plan_mutator(plan, msg_proc)
    120             ret = result_stack.pop()
    121             try:
--> 122                 msg = plan_stack[-1].send(ret)
    123             except StopIteration as e:
    124                 # discard the exhausted generator

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in plan_mutator(plan, msg_proc)
    167                     continue
    168                 else:
--> 169                     raise ex
    170         # if inserting / mutating, put new generator on the stack
    171         # and replace the current msg with the first element from the

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in plan_mutator(plan, msg_proc)
    120             ret = result_stack.pop()
    121             try:
--> 122                 msg = plan_stack[-1].send(ret)
    123             except StopIteration as e:
    124                 # discard the exhausted generator

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in fly_during_wrapper(plan, flyers)
    846     plan1 = plan_mutator(plan, insert_after_open)
    847     plan2 = plan_mutator(plan1, insert_before_close)
--> 848     return (yield from plan2)
    849 
    850 

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in plan_mutator(plan, msg_proc)
    167                     continue
    168                 else:
--> 169                     raise ex
    170         # if inserting / mutating, put new generator on the stack
    171         # and replace the current msg with the first element from the

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in plan_mutator(plan, msg_proc)
    120             ret = result_stack.pop()
    121             try:
--> 122                 msg = plan_stack[-1].send(ret)
    123             except StopIteration as e:
    124                 # discard the exhausted generator

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in plan_mutator(plan, msg_proc)
    167                     continue
    168                 else:
--> 169                     raise ex
    170         # if inserting / mutating, put new generator on the stack
    171         # and replace the current msg with the first element from the

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in plan_mutator(plan, msg_proc)
    120             ret = result_stack.pop()
    121             try:
--> 122                 msg = plan_stack[-1].send(ret)
    123             except StopIteration as e:
    124                 # discard the exhausted generator

/epics/xf/23id/ipython/profile_collection/startup/csx1/plans/nanop.py in spiral_continuous(detectors, x_motor, y_motor, x_start, y_start, npts, probe_size, overlap, tilt, per_step, md)
    110 
    111     try:
--> 112         return (yield from plan_steps())
    113 
    114     except Exception:

/epics/xf/23id/ipython/profile_collection/startup/csx1/plans/nanop.py in plan_steps()
    105 
    106     def plan_steps():
--> 107         yield from cont_sp_plan
    108         print('Moving back to first point position.')
    109         yield from reset_plan

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/plans.py in scan_nd(detectors, cycler, per_step, md)
    748             yield from per_step(detectors, step, pos_cache)
    749 
--> 750     return (yield from inner_scan_nd())
    751 
    752 

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/utils.py in dec_inner(*inner_args, **inner_kwargs)
   1009                 plan = gen_func(*inner_args, **inner_kwargs)
   1010                 plan = wrapper(plan, *args, **kwargs)
-> 1011                 return (yield from plan)
   1012             return dec_inner
   1013         return dec

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in stage_wrapper(plan, devices)
    937         return (yield from plan)
    938 
--> 939     return (yield from finalize_wrapper(inner(), unstage_devices()))
    940 
    941 

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in finalize_wrapper(plan, final_plan, pause_for_debug)
    498     cleanup = True
    499     try:
--> 500         ret = yield from plan
    501     except GeneratorExit:
    502         cleanup = False

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in inner()
    935     def inner():
    936         yield from stage_devices()
--> 937         return (yield from plan)
    938 
    939     return (yield from finalize_wrapper(inner(), unstage_devices()))

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/utils.py in dec_inner(*inner_args, **inner_kwargs)
   1009                 plan = gen_func(*inner_args, **inner_kwargs)
   1010                 plan = wrapper(plan, *args, **kwargs)
-> 1011                 return (yield from plan)
   1012             return dec_inner
   1013         return dec

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in run_wrapper(plan, md)
    324     yield from contingency_wrapper(plan,
    325                                    except_plan=except_plan,
--> 326                                    else_plan=close_run)
    327     return rs_uid
    328 

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in contingency_wrapper(plan, except_plan, else_plan, final_plan, pause_for_debug)
    560     cleanup = True
    561     try:
--> 562         ret = yield from plan
    563     except GeneratorExit:
    564         cleanup = False

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/plans.py in inner_scan_nd()
    746     def inner_scan_nd():
    747         for step in list(cycler):
--> 748             yield from per_step(detectors, step, pos_cache)
    749 
    750     return (yield from inner_scan_nd())

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/plan_stubs.py in one_nd_step(detectors, step, pos_cache)
    950     motors = step.keys()
    951     yield from move_per_step(step, pos_cache)
--> 952     yield from trigger_and_read(list(detectors) + list(motors))
    953 
    954 

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/plan_stubs.py in trigger_and_read(devices, name)
    800     from .preprocessors import rewindable_wrapper
    801     return (yield from rewindable_wrapper(inner_trigger_and_read(),
--> 802                                           rewindable))
    803 
    804 

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in rewindable_wrapper(plan, rewindable)
    678                                             restore_rewindable()))
    679     else:
--> 680         return (yield from plan)
    681 
    682 

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/plan_stubs.py in inner_trigger_and_read()
    796             if reading is not None:
    797                 ret.update(reading)
--> 798         yield from save()
    799         return ret
    800     from .preprocessors import rewindable_wrapper

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/plan_stubs.py in save()
     55     :func:`bluesky.plan_stubs.create`
     56     """
---> 57     return (yield Msg('save'))
     58 
     59 

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in plan_mutator(plan, msg_proc)
    192         try:
    193             # yield out the 'current message' and collect the return
--> 194             inner_ret = yield msg
    195         except GeneratorExit:
    196             # special case GeneratorExit.  We must clean up all of our plans

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in plan_mutator(plan, msg_proc)
    192         try:
    193             # yield out the 'current message' and collect the return
--> 194             inner_ret = yield msg
    195         except GeneratorExit:
    196             # special case GeneratorExit.  We must clean up all of our plans

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in plan_mutator(plan, msg_proc)
    192         try:
    193             # yield out the 'current message' and collect the return
--> 194             inner_ret = yield msg
    195         except GeneratorExit:
    196             # special case GeneratorExit.  We must clean up all of our plans

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in plan_mutator(plan, msg_proc)
    192         try:
    193             # yield out the 'current message' and collect the return
--> 194             inner_ret = yield msg
    195         except GeneratorExit:
    196             # special case GeneratorExit.  We must clean up all of our plans

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in plan_mutator(plan, msg_proc)
    192         try:
    193             # yield out the 'current message' and collect the return
--> 194             inner_ret = yield msg
    195         except GeneratorExit:
    196             # special case GeneratorExit.  We must clean up all of our plans

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/run_engine.py in _run(self)
   1209                         # exceptions (coming in via throw) can be
   1210                         # raised
-> 1211                         new_response = yield from coro(msg)
   1212 
   1213                     # special case `CancelledError` and let the outer

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/run_engine.py in _save(self, msg)
   1685                        configuration=config, name=desc_key,
   1686                        hints=hints, object_keys=object_keys)
-> 1687             yield from self.emit(DocumentNames.descriptor, doc)
   1688             self.log.debug("Emitted Event Descriptor with name %r containing "
   1689                            "data keys %r (uid=%r)", desc_key,

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/asyncio/coroutines.py in coro(*args, **kw)
    210         @functools.wraps(func)
    211         def coro(*args, **kw):
--> 212             res = func(*args, **kw)
    213             if (base_futures.isfuture(res) or inspect.isgenerator(res) or
    214                 isinstance(res, CoroWrapper)):

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/run_engine.py in emit(self, name, doc)
   2322         "Process blocking callbacks and schedule non-blocking callbacks."
   2323         _validate(doc, schemas[name])
-> 2324         self.dispatcher.process(name, doc)
   2325 
   2326 

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/run_engine.py in process(self, name, doc)
   2342         doc : dict
   2343         """
-> 2344         exceptions = self.cb_registry.process(name, name.name, doc)
   2345         for exc, traceback in exceptions:
   2346             warn("A %r was raised during the processing of a %s "

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/utils.py in process(self, sig, *args, **kwargs)
    332             for cid, func in list(self.callbacks[sig].items()):
    333                 try:
--> 334                     func(*args, **kwargs)
    335                 except ReferenceError:
    336                     self._remove_proxy(func)

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/utils.py in __call__(self, *args, **kwargs)
    413             mtd = self.func
    414         # invoke the callable and return the result
--> 415         return mtd(*args, **kwargs)
    416 
    417     def __eq__(self, other):

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/callbacks/best_effort.py in __call__(self, name, doc)
     91             return
     92 
---> 93         super().__call__(name, doc)
     94 
     95     def start(self, doc):

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/callbacks/core.py in __call__(self, name, doc)
     15     def __call__(self, name, doc):
     16         "Dispatch to methods expecting particular doc types."
---> 17         return getattr(self, name)(doc)
     18 
     19     def event(self, doc):

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/callbacks/best_effort.py in descriptor(self, doc)
    339                                                # Let clim autoscale.
    340                                                ax=ax)
--> 341                     live_scatter('start', self._start_doc)
    342                     live_scatter('descriptor', doc)
    343                     self._live_scatters[doc['uid']][I_key] = live_scatter

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/callbacks/core.py in __call__(self, name, doc)
     15     def __call__(self, name, doc):
     16         "Dispatch to methods expecting particular doc types."
---> 17         return getattr(self, name)(doc)
     18 
     19     def event(self, doc):

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/callbacks/mpl_plotting.py in start(self, doc)
    223         self._Idata.clear()
    224         sc = self.ax.scatter(self._xdata, self._ydata, c=self._Idata,
--> 225                              norm=self._norm, cmap=self.cmap, **self.kwargs)
    226         self._sc.append(sc)
    227         self.sc = sc

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/matplotlib/__init__.py in inner(ax, data, *args, **kwargs)
   1803                         "the Matplotlib list!)" % (label_namer, func.__name__),
   1804                         RuntimeWarning, stacklevel=2)
-> 1805             return func(ax, *args, **kwargs)
   1806 
   1807         inner.__doc__ = _add_data_doc(inner.__doc__,

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/matplotlib/axes/_axes.py in scatter(self, x, y, s, c, marker, cmap, norm, vmin, vmax, alpha, linewidths, verts, edgecolors, **kwargs)
   4193                 isinstance(c, str) or
   4194                 (isinstance(c, collections.Iterable) and
-> 4195                     isinstance(c[0], str))):
   4196             c_array = None
   4197         else:

IndexError: list index out of range

''''

This one looks like a problem we had in the past and I thought it was closed.. 

''''
Gain bit set to 1 for a gain value of x2

Transient Scan ID: 114375     Time: 2019/03/11 05:15:26
Persistent Unique Scan ID: 'f56df647-316c-4c8d-96cc-bf3673a4aa3c'
New stream: 'baseline'
New stream: 'primary'                                                                                                                                                                               
+-----------+------------+-------------------+-------------------+-------------------+-------------------+-------------------+
|   seq_num |       time | fccd_stats1_total | fccd_stats2_total | fccd_stats3_total | fccd_stats4_total | fccd_stats5_total |
+-----------+------------+-------------------+-------------------+-------------------+-------------------+-------------------+
|         1 | 05:17:31.5 |         -81436250 |       -1519153877 |       -1577582637 |      -10494122898 |      -30097444501 |
+-----------+------------+-------------------+-------------------+-------------------+-------------------+-------------------+
generator count ['f56df647'] (scan num: 114375)


Gain bit set to 2 for a gain value of x1

Transient Scan ID: 114376     Time: 2019/03/11 05:17:32
Persistent Unique Scan ID: 'bc250cf7-eec2-4a33-9d44-ed7d8111bb2d'
New stream: 'baseline'
New stream: 'primary'                                                                                                                                                                               
+-----------+------------+-------------------+-------------------+-------------------+-------------------+-------------------+
|   seq_num |       time | fccd_stats1_total | fccd_stats2_total | fccd_stats3_total | fccd_stats4_total | fccd_stats5_total |
+-----------+------------+-------------------+-------------------+-------------------+-------------------+-------------------+
|         1 | 05:19:37.3 |         -40376192 |        -753122714 |        -782820638 |       -5206202008 |      -14962529993 |
+-----------+------------+-------------------+-------------------+-------------------+-------------------+-------------------+
generator count ['bc250cf7'] (scan num: 114376)

Returning to intial conditions (pre-count).
        Total images per trigger are NOW:        120                                                                                                                                                
	FCCD FCRIC gain value is NOW:		 auto


[E 05:19:39.666 run_engine:1264] Run aborted
    Traceback (most recent call last):
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/run_engine.py", line 1130, in _run
        self._exception or resp)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 1294, in __call__
        return (yield from plan)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 1149, in baseline_wrapper
        return (yield from plan_mutator(plan, insert_baseline))
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 77, in plan_mutator
        msg = plan_stack[-1].throw(exception)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 790, in monitor_during_wrapper
        return (yield from plan2)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 77, in plan_mutator
        msg = plan_stack[-1].throw(exception)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 77, in plan_mutator
        msg = plan_stack[-1].throw(exception)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 848, in fly_during_wrapper
        return (yield from plan2)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 77, in plan_mutator
        msg = plan_stack[-1].throw(exception)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 77, in plan_mutator
        msg = plan_stack[-1].throw(exception)
      File "/home/xf23id1/Beamline/ScienceComm/2019_03_ReconsTeam/Ptychography_routine.py", line 51, in night_20190311
        yield from mv(fccd.exposure,(0.5,1,5))
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/plan_stubs.py", line 245, in mv
        ret = yield Msg('set', obj, val, group=group)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 194, in plan_mutator
        inner_ret = yield msg
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 194, in plan_mutator
        inner_ret = yield msg
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 194, in plan_mutator
        inner_ret = yield msg
      [Previous line repeated 1 more times]
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/run_engine.py", line 1211, in _run
        new_response = yield from coro(msg)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/asyncio/coroutines.py", line 212, in coro
        res = func(*args, **kw)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/run_engine.py", line 1958, in _set
        ret = msg.obj.set(*msg.args, **kwargs)
      File "/epics/xf/23id/ipython/profile_collection/startup/csx1/devices/areadetector.py", line 219, in set
        self.parent.cam.num_images.set(exp[2])
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/ophyd/signal.py", line 941, in set
        return super().set(value, timeout=timeout, settle_time=settle_time)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/ophyd/signal.py", line 185, in set
        raise RuntimeError('Another set() call is still in progress')
    RuntimeError: Another set() call is still in progress
[E 05:19:39.666 run_engine:1264] Run aborted
    Traceback (most recent call last):
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/run_engine.py", line 1130, in _run
        self._exception or resp)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 1294, in __call__
        return (yield from plan)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 1149, in baseline_wrapper
        return (yield from plan_mutator(plan, insert_baseline))
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 77, in plan_mutator
        msg = plan_stack[-1].throw(exception)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 790, in monitor_during_wrapper
        return (yield from plan2)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 77, in plan_mutator
        msg = plan_stack[-1].throw(exception)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 77, in plan_mutator
        msg = plan_stack[-1].throw(exception)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 848, in fly_during_wrapper
        return (yield from plan2)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 77, in plan_mutator
        msg = plan_stack[-1].throw(exception)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 77, in plan_mutator
        msg = plan_stack[-1].throw(exception)
      File "/home/xf23id1/Beamline/ScienceComm/2019_03_ReconsTeam/Ptychography_routine.py", line 51, in night_20190311
        yield from mv(fccd.exposure,(0.5,1,5))
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/plan_stubs.py", line 245, in mv
        ret = yield Msg('set', obj, val, group=group)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 194, in plan_mutator
        inner_ret = yield msg
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 194, in plan_mutator
        inner_ret = yield msg
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py", line 194, in plan_mutator
        inner_ret = yield msg
      [Previous line repeated 1 more times]
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/run_engine.py", line 1211, in _run
        new_response = yield from coro(msg)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/asyncio/coroutines.py", line 212, in coro
        res = func(*args, **kw)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/run_engine.py", line 1958, in _set
        ret = msg.obj.set(*msg.args, **kwargs)
      File "/epics/xf/23id/ipython/profile_collection/startup/csx1/devices/areadetector.py", line 219, in set
        self.parent.cam.num_images.set(exp[2])
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/ophyd/signal.py", line 941, in set
        return super().set(value, timeout=timeout, settle_time=settle_time)
      File "/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/ophyd/signal.py", line 185, in set
        raise RuntimeError('Another set() call is still in progress')
    RuntimeError: Another set() call is still in progress
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
~/Beamline/ScienceComm/2019_03_ReconsTeam/Ptychography_routine.py in <module>
----> 1 RE(night_20190311())

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/run_engine.py in __call__(self, *args, **metadata_kw)
    754                     # it (unless it is a canceled error)
    755                     if exc is not None:
--> 756                         raise exc
    757 
    758             if self._interrupted:

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/run_engine.py in _run(self)
   1263             self._reason = str(err)
   1264             self.log.exception("Run aborted")
-> 1265             raise err
   1266         finally:
   1267             # Some done_callbacks may still be alive in other threads.

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/run_engine.py in _run(self)
   1128                         try:
   1129                             msg = self._plan_stack[-1].throw(
-> 1130                                 self._exception or resp)
   1131                         except Exception as e:
   1132                             # The current plan did not handle it,

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in __call__(self, plan)
   1292         plan = monitor_during_wrapper(plan, self.monitors)
   1293         plan = baseline_wrapper(plan, self.baseline)
-> 1294         return (yield from plan)

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in baseline_wrapper(plan, devices, name)
   1147         return (yield from plan)
   1148     else:
-> 1149         return (yield from plan_mutator(plan, insert_baseline))
   1150 
   1151 

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in plan_mutator(plan, msg_proc)
     75             # if we have a stashed exception, pass it along
     76             try:
---> 77                 msg = plan_stack[-1].throw(exception)
     78             except StopIteration as e:
     79                 # discard the exhausted generator

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in monitor_during_wrapper(plan, signals)
    788     plan1 = plan_mutator(plan, insert_after_open)
    789     plan2 = plan_mutator(plan1, insert_before_close)
--> 790     return (yield from plan2)
    791 
    792 

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in plan_mutator(plan, msg_proc)
     75             # if we have a stashed exception, pass it along
     76             try:
---> 77                 msg = plan_stack[-1].throw(exception)
     78             except StopIteration as e:
     79                 # discard the exhausted generator

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in plan_mutator(plan, msg_proc)
     75             # if we have a stashed exception, pass it along
     76             try:
---> 77                 msg = plan_stack[-1].throw(exception)
     78             except StopIteration as e:
     79                 # discard the exhausted generator

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in fly_during_wrapper(plan, flyers)
    846     plan1 = plan_mutator(plan, insert_after_open)
    847     plan2 = plan_mutator(plan1, insert_before_close)
--> 848     return (yield from plan2)
    849 
    850 

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in plan_mutator(plan, msg_proc)
     75             # if we have a stashed exception, pass it along
     76             try:
---> 77                 msg = plan_stack[-1].throw(exception)
     78             except StopIteration as e:
     79                 # discard the exhausted generator

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in plan_mutator(plan, msg_proc)
     75             # if we have a stashed exception, pass it along
     76             try:
---> 77                 msg = plan_stack[-1].throw(exception)
     78             except StopIteration as e:
     79                 # discard the exhausted generator

~/Beamline/ScienceComm/2019_03_ReconsTeam/Ptychography_routine.py in night_20190311()
     49     step_num=41
     50     #olog('Square mesh #{} @ {:.3f} eV, span 0.02 mm, steps {:d}'.format((db[-1].get('start').get('scan_id'))+1, pgm.energy.readback.value, int(step_num)))
---> 51     yield from mv(fccd.exposure,(0.5,1,5))
     52     yield from relative_outer_product_scan(dets,nanop.bz,-0.01,0.01,step_num,nanop.bx,-0.01,0.01,step_num,0)
     53     yield from mv(fccd.exposure,(0.5,1,120))

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/plan_stubs.py in mv(*args)
    243     step, = utils.merge_cycler(cyl)
    244     for obj, val in step.items():
--> 245         ret = yield Msg('set', obj, val, group=group)
    246         status_objects.append(ret)
    247     yield Msg('wait', None, group=group)

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in plan_mutator(plan, msg_proc)
    192         try:
    193             # yield out the 'current message' and collect the return
--> 194             inner_ret = yield msg
    195         except GeneratorExit:
    196             # special case GeneratorExit.  We must clean up all of our plans

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in plan_mutator(plan, msg_proc)
    192         try:
    193             # yield out the 'current message' and collect the return
--> 194             inner_ret = yield msg
    195         except GeneratorExit:
    196             # special case GeneratorExit.  We must clean up all of our plans

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in plan_mutator(plan, msg_proc)
    192         try:
    193             # yield out the 'current message' and collect the return
--> 194             inner_ret = yield msg
    195         except GeneratorExit:
    196             # special case GeneratorExit.  We must clean up all of our plans

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in plan_mutator(plan, msg_proc)
    192         try:
    193             # yield out the 'current message' and collect the return
--> 194             inner_ret = yield msg
    195         except GeneratorExit:
    196             # special case GeneratorExit.  We must clean up all of our plans

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/preprocessors.py in plan_mutator(plan, msg_proc)
    192         try:
    193             # yield out the 'current message' and collect the return
--> 194             inner_ret = yield msg
    195         except GeneratorExit:
    196             # special case GeneratorExit.  We must clean up all of our plans

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/run_engine.py in _run(self)
   1209                         # exceptions (coming in via throw) can be
   1210                         # raised
-> 1211                         new_response = yield from coro(msg)
   1212 
   1213                     # special case `CancelledError` and let the outer

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/asyncio/coroutines.py in coro(*args, **kw)
    210         @functools.wraps(func)
    211         def coro(*args, **kw):
--> 212             res = func(*args, **kw)
    213             if (base_futures.isfuture(res) or inspect.isgenerator(res) or
    214                 isinstance(res, CoroWrapper)):

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/run_engine.py in _set(self, msg)
   1956         group = kwargs.pop('group', None)
   1957         self._movable_objs_touched.add(msg.obj)
-> 1958         ret = msg.obj.set(*msg.args, **kwargs)
   1959         p_event = asyncio.Event(loop=self.loop)
   1960         pardon_failures = self._pardon_failures

/epics/xf/23id/ipython/profile_collection/startup/csx1/devices/areadetector.py in set(self, exp)
    217 
    218         if exp[2] is not None:
--> 219             self.parent.cam.num_images.set(exp[2])
    220 
    221         return NullStatus()

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/ophyd/signal.py in set(self, value, timeout, settle_time)
    939         '''
    940         if not self._put_complete:
--> 941             return super().set(value, timeout=timeout, settle_time=settle_time)
    942 
    943         # using put completion:

/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/ophyd/signal.py in set(self, value, timeout, settle_time)
    183 
    184         if self._set_thread is not None:
--> 185             raise RuntimeError('Another set() call is still in progress')
    186 
    187         st = Status(self)

RuntimeError: Another set() call is still in progress

In [37]:                                                                                                                                                                                            

```
''''

**Comment by danielballan** [comment]
Looking at that final error,

```
RuntimeError: Another set() call is still in progress
```

I think this is a duplicate of #1102, which concluded as:

> This is not bluesky bug, it is either in the beamline IOC or in the beamline profile, closing.

Is this the same issue?

**Comment by prjemian** [comment]
I get this occasionally.  Usually a symptom of using a blocking call:
    signal.put(value)
in a plan, instead of using:
    yield from bps.mv(signal, value)

Might that be the case here?

**Comment by prjemian** [comment]
Caught the weasel, live, as shown:

```
/epics/xf/23id/ipython/profile_collection/startup/csx1/devices/areadetector.py in set(self, exp)
    217 
    218         if exp[2] is not None:
--> 219             self.parent.cam.num_images.set(exp[2])
    220 
    221         return NullStatus()
```

Perhaps you want to use:

```
    if exp[2] is not None:
        yield from bps.abs_set(self.parent.cam.num_images, exp[2])
```

**Comment by ambarb** [comment]
@cmazzoli was this during the dark image collection?  the trace back has other scans mixed in between.  If it is dark image collection, then I think @prjemian has given us something to check in the ophyd device or the dark plan.  I don't remember why we decided to use `bps.mv()`

if this issue you are reporting is not caused by dark collection with our dark plans, then I think there is a different problem.  Please see also 
https://github.com/NSLS-II-CSX/xf23id1_profiles/issues/25 
https://github.com/NSLS-II-CSX/xf23id1_profiles/issues/38

**Comment by prjemian** [comment]
`bps.mv()` waits for completion.  If you are getting *another `set()` in progress* with acquisitions very close together, it's a real indication that the process is not waiting for the previous acquisition to finish before starting a new one.  With EPICS area detector, need to ensure that all AD plugins have finished processing before starting the next acquire from the camera.

**Comment by prjemian** [comment]
FWIW: With the new ophyd release 1.6.0, these *another `set()` in progress* exceptions have not interrupted any of our USAXS scans (XPCS has not used it yet).  Before 1.6.0, this exception would raise within hours.  We now use this code in our [initial instrument setup](https://github.com/APS-USAXS/ipython-usaxs/blob/master/profile_bluesky/startup/instrument/framework/initialize.py):

```python
# set default timeout for all EpicsSignal connections & communications
from ophyd.signal import EpicsSignalBase

try:
    EpicsSignalBase.set_defaults(
        auto_monitor=True,
        timeout=60,
        write_timeout=60,
        connection_timeout=5,
    )
except Exception as exc:
    warnings.warn(
        "ophyd version is old, upgrade to 1.6.0+ "
        "to get set_defaults() method"
        )
    EpicsSignalBase.set_default_timeout(timeout=10, connection_timeout=5)
```

**Comment by ambarb** [comment]
@prjemian thanks.  I could not remember if we needed to wait for completion or not.  I think we did because we are making print statements to the screen.    I think our IOC has put complete capabilities so we were trying to use that. If the print statements are the only reason, then maybe we should just use `set()` .

Good news about `1.6.0`. I will check to see what we have at the beamline.  Thanks for the heads up.  Did you need to add the code snippet above to deal with pre-`1.6.0`or did `1.6.0` allow you to add the snippet?

**Comment by ambarb** [comment]
never mind. I read the exception message

**Comment by prjemian** [comment]
The `try..except` wrapping is for beamlines *in transition* to newer code.  They have the option to use a conda environment with ophyd <1.6.0.  Always, the beam lines want a fall-back position.

**Comment by stan-dot** [comment]
@prjemian  the current ophyd is 1.9.0 the issue might not be there anymore, I am not sure how to replicate though

**Comment by ambarb** [comment]
this maybe should have been closed - it depends on where you want to fix the issue

we have a `bps.mv(inout, "In")` that blocks the beam with something that is not the detectors.  it is related to the `TwoButtonShutter`-like ophyd devices.  Once the device was updated to try to move multiple times, we do not get this error any more.  Even if the plan fails (because the multiple tries are limited to 5 or so), it fails gracefully and does make the detector (used in the line after `bps.mv(inout, "In")` to be in a strange state inside the bluesky name space.

if you dig to find the PR for [this line](https://github.com/NSLS-II-CSX/profile_collection/blame/429cf3846eb42aecddfe5ad839c007a1b8f79f58/startup/csx1/devices/eps.py#L62), you will see all the changes Dan made to our profile

**Comment by stan-dot** [comment]
I guess since the specific bug was fixed in a different way and this error was with python3.6
``` 
/opt/conda_envs/collection-2019-1.2-csx/lib/python3.6/site-packages/bluesky/utils.py in ensure_generator(plan)
    110     if isinstance(plan, Msg):
    111         return single_gen(plan)
--> 112     gen = iter(plan)  # no-op on generators; needed for classes
    113     if not isinstance(gen, PLAN_TYPES):
    114         # If plan does not support .send, we must wrap it in a generator.

TypeError: 'NoneType' object is not iterable
```

Then I'd close the specific issue @ambarb

**Comment by physwkim** [comment]
At PLS-II, we frequently encounter the error message: ```Another set() call is still in progress``` across several beamlines. However, when we set ```some_signal.put_complete = True```, the error does not occur.

I suspect that although the set operation in EPICS has completed, the callback does not recognize this and continues to wait. When we use the ```put_complete = True``` option, the error does not occur, as Bluesky does not rely on the callback in this case.

**Comment by ambarb** [comment]
yes, #1102 is related and all of this was fixed some time ago in the beamline's profile.  The  TwoButtonShutter-like ophyd device was the cause of this, but ours was a customized version that can work with things that are not just "Opn" or "Cls" (open or closed). 

A newer version of bluesky had improved tracebacks to point us to the thing that was actually causing the problem.  It had nothing to do with the CCD.   I don't believe i am able to close the issue, but I will try now.

https://github.com/NSLS-II-CSX/profile_collection/pull/51
with and edge case improvment with this
https://github.com/NSLS-II-CSX/profile_collection/pull/73
